pkg/agent/openflow/feature.go

```diff
@@ -0,0 +1,280 @@
+//            |       - table C        |                              |        - table G        |
+//            +------------------------+                              +-------------------------+
+
+// feature is the interface to program a major function in Antrea data path. The following structures implement this interface:
+//
+// - featurePodConnectivity, implementation of connectivity for Pods, activated by default.
+// - featureNetworkPolicy, implementation of K8s NetworkPolicy and Antrea NetworkPolicy, activated by default.
+// - featureService, implementation of K8s Service, activated by default.
+// - featureEgress, implementation of Egress, activation is determined by feature gate Egress.
+// - featureMulticast, implementation of multicast, activation is determined by feature gate Multicast,
+// - featureTraceflow, implementation of Traceflow.
+type feature interface {
+	// getFeatureName returns the name of the feature.
+	getFeatureName() string
+	// getRequiredTables returns a slice of required tables of the feature.
+	getRequiredTables() []*Table
+	// initFlows returns the initial flows of the feature.
+	initFlows() []binding.Flow
+	// replayFlows returns the fixed and cached flows that need to be replayed after OVS is reconnected.
+	replayFlows() []binding.Flow
+}
+
+const (
+	// Pipeline is used to implement a major function in Antrea data path. At this moment, we have the following pipelines:
+
+	// pipelineARP is used to process ARP packets.
+	pipelineARP binding.PipelineID = iota
+	// pipelineIP is used to process IPv4/IPv6 packets
+	pipelineIP
+	// pipelineMulticast is used to process multicast packets.
+	pipelineMulticast
+	// pipelineAll is only used for PipelineClassifierTable that doesn't belong to any pipelines.
+	pipelineAll binding.PipelineID = 255
+)
+
+const (
+	// Stage is used to group tables which implement similar functions in a pipeline. At this moment, we have the following
+	// stages:
+
+	// Classify packets "category" (tunnel, local gateway or local Pod, etc).
+	stageClassifier binding.StageID = iota
+	// Validate packets.
+	stageValidation
+	// Transform committed packets in CT zones.
+	stageConntrackState
+	// Similar to PREROUTING chain of nat table in iptables. DNAT for Service connections is performed in this stage.
+	stagePreRouting
+	// Install egress rules for K8s NetworkPolicy and Antrea NetworkPolicy.
+	stageEgressSecurity
+	// L3 Forwarding of packets.
+	stageRouting
+	// Similar to POSTROUTING chain of nat table in iptables. SNAT for Service connections is performed in this stage.
+	stagePostRouting
+	// L2 Forwarding of packets.
+	stageSwitching
+	// Install ingress rules for K8s NetworkPolicy and Antrea NetworkPolicy.
+	stageIngressSecurity
+	// Commit non-Service connections.
+	stageConntrack
+	// Output packets to target port.
+	stageOutput
+	// stageAll is only used for PipelineClassifierTable that doesn't belong to any stages.
```

Comment:
But it indeed sounds like the first stage? Should we use 0, and name it sth like stage0，stageStart, etc.?

Reply:
I think stageStart is ok.

Reply:
I think stageStart is ok.

Reply:
Thanks for reviewing this PR!

Reply:
Thanks for reviewing this PR!